### PR TITLE
CMake: Enable -Werror=unused-parameter flag to treat unused-parameter warnings as errors

### PR DESCRIPTION
**Pull request by VasuDevrani**
for #2266 
this will output `unused-parameters` warnings as errors.

Screenshot 2024-05-15 133023

### CODE REVIEW

**Comment by torwig** [comment]
@VasuDevrani Should we also fix those errors in the scope of this PR?

**Comment by VasuDevrani** [comment]
> @VasuDevrani Should we also fix those errors in the scope of this PR?

maybe yes (but they are a lot), I think the CI failing because of those.

**Comment by PragmaTwice** [comment]
Yeah we should fix them otherwise the PR will be blocked by CI failure.

You can use the attribute `[[maybe_unused]]`.

**Comment by VasuDevrani** [comment]
> You can use the attribute `[[maybe_unused]]`.

Wouldn't it be better to drop the unused parameter and keep the type only
```c
virtual Status ToNumber(int64_t*) const { return {Status::NotOK, "not supported"}; }
virtual Status ToBool(bool*) const { return {Status::NotOK, "not supported"}; }
```

**Comment by mapleFU** [comment]
>  Wouldn't it be better to drop the unused parameter and keep the type only

It's also ok here but keeping a name might be more specific?

**Comment by PragmaTwice** [comment]
> > You can use the attribute `[[maybe_unused]]`.
> 
> Wouldn't it be better to drop the unused parameter and keep the type only
> 
> ```c
> virtual Status ToNumber(int64_t*) const { return {Status::NotOK, "not supported"}; }
> virtual Status ToBool(bool*) const { return {Status::NotOK, "not supported"}; }
> ```

There's some information in the names, so we'd better to keep them.

**Comment by VasuDevrani** [comment]
Two more concerns, then I'll move forward with `[[maybe_unused]]`:
- it can make code a bit unreadable (especially if there's multiple unused params together)
```c
void performOperation(const Config& config, [[maybe_unused]] int param1, [[maybe_unused]] double param2, [[maybe_unused]] const std::string& param3) {
    // Use only the 'config' parameter
    if (config.isEnabled()) {
        // Perform the operation
        // ...
    }
}
```
VS
```c
void performOperation(const Config& config, int, double, const std::string&) {
    // Use only the 'config' parameter
    if (config.isEnabled()) {
        // Perform the operation
        // ...
    }
}
```
- supported for cpp 17+ (then again project cmake sets it at 17)

**Comment by mapleFU** [comment]
We already uses C++17, yeah?

**Comment by VasuDevrani** [comment]
> We already uses C++17, isn't we?

yes

File: CMakeLists.txt [hunk 1]
```diff
@@ -273,6 +273,8 @@ else()
     target_compile_definitions(kvrocks_objs PUBLIC METADATA_ENCODING_VERSION=0)
 endif()
 
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Werror=unused-parameter")
```
Comment (mkmkme) [inline CMakeLists.txt:270]: Setting `CMAKE_CXX_FLAGS` globally is an anti-pattern in CMake.
Maybe the same effect can be achieved with 
```cmake
target_compile_options(kvrocks_objs PUBLIC -Werror=unused-parameter)
```

**Pull request closed without merging**